From 8dbaaf7174c6adf299cc2db322d4ce65806fa7f7 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Tue, 28 Jul 2026 00:31:35 -0400
Subject: [PATCH 241/268] winewayland: Remove pointer relative mode.

Windows always tracks an absolute pointer position.
---
 dlls/winewayland.drv/wayland_pointer.c | 34 ++++----------------------
 dlls/winewayland.drv/waylanddrv.h      |  1 -
 2 files changed, 5 insertions(+), 30 deletions(-)

diff --git a/dlls/winewayland.drv/wayland_pointer.c b/dlls/winewayland.drv/wayland_pointer.c
index 36dc5e45042..49e22afcfe4 100644
--- a/dlls/winewayland.drv/wayland_pointer.c
+++ b/dlls/winewayland.drv/wayland_pointer.c
@@ -182,8 +182,8 @@ static void pointer_handle_motion(void *data, struct wl_pointer *wl_pointer,
 {
     struct wayland_pointer *pointer = &process_wayland.pointer;
 
-    /* Ignore absolute motion events if in relative mode. */
-    if (pointer->relative_mode) return;
+    /* If the pointer is locked then it may not respect the win32u position */
+    if (pointer->zwp_locked_pointer_v1) return;
 
     pointer_handle_motion_internal(sx, sy);
 }
@@ -999,7 +999,7 @@ static void wayland_pointer_update_constraint(struct wl_surface *wl_surface,
                                               BOOL force_lock)
 {
     struct wayland_pointer *pointer = &process_wayland.pointer;
-    BOOL needs_relative, needs_lock, needs_confine, is_visible;
+    BOOL needs_lock, needs_confine, is_visible;
     static unsigned int once;
 
     if (!process_wayland.zwp_pointer_constraints_v1)
@@ -1089,26 +1089,7 @@ static void wayland_pointer_update_constraint(struct wl_surface *wl_surface,
         }
     }
 
-    if (!process_wayland.zwp_relative_pointer_manager_v1)
-    {
-        if (!once++)
-            ERR("zwp_relative_pointer_manager_v1 isn't supported, skipping relative motion\n");
-        return;
-    }
-
-    needs_relative = !is_visible && pointer->constraint_hwnd &&
-                     pointer->constraint_hwnd == pointer->focused_hwnd;
-
-    if (needs_relative)
-    {
-        pointer->relative_mode = TRUE;
-        TRACE("Enabling relative motion\n");
-    }
-    else if (!needs_relative)
-    {
-        pointer->relative_mode = FALSE;
-        TRACE("Disabling relative motion\n");
-    }
+    TRACE("lock=%u confine=%u\n", needs_lock, needs_confine);
 }
 
 void wayland_pointer_clear_constraint(void)
@@ -1134,11 +1115,6 @@ BOOL WAYLAND_SetCursorPos(INT x, INT y)
     struct wayland_pointer *pointer = &process_wayland.pointer;
 
     pthread_mutex_lock(&pointer->mutex);
-    if (pointer->relative_mode)
-    {
-        pthread_mutex_unlock(&pointer->mutex);
-        return FALSE;
-    }
     pointer->pending_warp = TRUE;
     pointer->warp.x = x;
     pointer->warp.y = y;
@@ -1196,7 +1172,7 @@ BOOL WAYLAND_ClipCursor(const RECT *clip, BOOL reset)
                     wl_fixed_from_int(warp.y),
                     pointer->enter_serial);
             TRACE("warp_pointer hwnd=%p wayland_xy=%s screen_xy=%s\n",
-                    hwnd, wine_dbgstr_point(&warp), wine_dbgstr_point(&cursor_pos));
+                  hwnd, wine_dbgstr_point(&warp), wine_dbgstr_point(&cursor_pos));
         }
         else
         {
diff --git a/dlls/winewayland.drv/waylanddrv.h b/dlls/winewayland.drv/waylanddrv.h
index 0b4f2be6357..9f7501723bd 100644
--- a/dlls/winewayland.drv/waylanddrv.h
+++ b/dlls/winewayland.drv/waylanddrv.h
@@ -161,7 +161,6 @@ struct wayland_pointer
     HWND constraint_hwnd;
     BOOL pending_warp;
     POINT warp;
-    BOOL relative_mode;
     uint32_t enter_serial;
     uint32_t button_serial;
     struct wayland_cursor cursor;
-- 
2.55.0

